Triggers for geometry

ichfarbstift opened the pull request:
Added 'area of' and 'perimeter of' to the triggers for the geometry IA, as of issue #2386.

A potential issue: queries like 'perimeter of circle' now result in the IA showing both circumference and area. Maybe this is not necessarily a bad thing, since the query 'perimeter of circle' or 'perimeter of sphere' are ambiguous queries. I am not sure whether more changes in this regard are necessary.

https://duck.co/ia/view/geometry

daxtheduck commented:
@ichfarbstift Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

While the team and the community review your submission, we'd like to make sure you look good! If you'd like a free DuckDuckGo tshirt, please send over the following info to merch@duckduckgo.com, and be sure to put "Dev Merch" in the subject line:

Name:
Street Address:
City:
State / Province:
Country:
Phone (Non-US Orders only): 
T-shirt size:
GitHub username:
Link to this Pull Request:

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please send an email to <quackslack@duckduckgo.com>.

>More Info: https://duck.co/duckduckhack/submission_and_review

Thanks!

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

GuiltyDolphin commented:
@ichfarbstift I see you found something, that's great!

Mind adding a few tests to `t/Geometry.t` to make sure the new triggers work?

pecey commented:
@moollaza Still open?

GuiltyDolphin commented:
@ichfarbstift How's this going?

@russellholt You happy with the changes so far?

GuiltyDolphin commented on t/Geometry.t:
```diff
@@ -21,6 +21,16 @@ ddg_spice_test(
+    'perimeter of square' => test_spice(
+        '/js/spice/geometry/',
+        call_type => 'self',
+        caller => 'DDG::Spice::Geometry'
```
Can we get a bunch of tests for things that shouldn't trigger this? Like `area of a cone` etc...

Would be nice to have some more valid tests as well.

ichfarbstift replied:
Hallo, sorry for not replying sooner. I am a bit busy at the moment, but I am planning to continue working on the tests next Tuesday.

GuiltyDolphin replied:
@ichfarbstift No worries! Just go at your own pace - there's no rush.

ichfarbstift commented:
I modified this spice a little bit. Now the triggers are either specific names for formulas (e.g., 'volume' or 'area') or generic ('geometry' and 'formulas'). Using a specific trigger the instant answer is displayed only if the shape requested indeed contains the specific formula; for example, 'circumference circle' works but 'circumference sphere' does not. Using the generic triggers all the formulas of the requested shape are displayed. Specific triggers have priority over generic ones. That is, the query 'geometry perimeter of square' only shows the formula for the perimeter.

I also introduced more test cases, some for the generic triggers, some for the specific triggers and some for which the spice fails.

GuiltyDolphin commented on share/spice/geometry/geometry.js:
```diff
@@ -448,7 +451,13 @@
-
+        //fails if the search string asks for a formula that does not belong to the shape,
+        //unless the query is triggered by the generic keywords 'geometry' and 'formulas'
+        if(found_formula == false && !(query.match("geometry") || query.match("formulas"))){
```
This doesn't feel like it will scale well with more triggers - perhaps we should move it out to an Array that we can then check against.

ichfarbstift commented:
Introduced a function isGeneric that returns true if the search text does not contain a formula that is specified in the variable formulas. In this way, if more triggers will be considered generic until a corresponding formula is defined in the code.

GuiltyDolphin commented on t/Geometry.t:
```diff
@@ -7,17 +7,72 @@ use DDG::Test::Spice;
         caller => 'DDG::Spice::Geometry'
     ),
-    'area rectangle formula' => test_spice(
+    'formulas square' => test_spice(
```
Lets add some tests for `rectangle area formula`, `formula for area of a sphere`, `sphere equations` etc...

ichfarbstift replied:
@GuiltyDolphin I would do that, but I am not sure how to write test files for tests that are not supposed to trigger the spice. Can you please point me to some page where I can read about it?
Thanks!

ichfarbstift replied:
@GuiltyDolphin Ok, I think I have found out how!

jagtalon commented on share/spice/geometry/geometry.js:
```diff
@@ -92,6 +92,13 @@
+        for(var key in formulas)
+            if(query.match(key))
+                return false;
+        return true;
```
Nice! Nitpick: please add braces in the for block and the if block

jagtalon commented on share/spice/geometry/geometry.js:
```diff
@@ -448,7 +458,12 @@
+        //if the query is not generic fails if search string asks for a formula that does not belong to the shape
+        if(found_formula == false && isGeneric(query) == false){
+            Spice.failed("geometry");
+            return;
```
You could do: `return Spice.failed('geometry')`

moollaza commented:
@ichfarbstift thanks again for all the hard work you've put into improving this Geometry Goodie. You've done a great job so far! We at DuckDuckGo, and the DuckDuckHack community really appreciate you taking the time to contribute.

It looks like this is just about finished? I'll get this deployed to Beta for testing shortly. If you're stuck or need any help, I'd love to chat if you're around this week or next. Please message me on slack (@moollaza) or send me an email (moollaza@duckduckgo.com) if you'd be interested in talking. 

Thanks again!

moollaza commented:
@ichfarbstift BTW the changes are now on beta..

moollaza commented:
This seems good to me?

It looks like the height of the square isn't as great because the content isn't as tall -- we may want to set a minimum height. As well, there aren't any labels for the sides so `A = a^2` is hard to understand because `a` isn't labelled...

moollaza commented:
Though I don't like comparing, Google does a great job at handling thing. They go one step further and allow you to enter a value for `a` and it calculates the area.

GuiltyDolphin commented:
@ichfarbstift Lets update the `name` field to `Math` (in the structured answer).

jdorweiler commented:
This pull request is being closed due to inactivity but, if anyone would like to continue it, please @ mention 'duckduckhack-contributors' for help and http://docs.duckduckhack.com/submitting/submitting-overview.html for reference.
